src/flight.rs

```diff
@@ -193,3 +202,204 @@ impl SqlFlightClient {
         Ok(stream)
     }
 }
+
+pub fn query_to_stream_with_params(
+    client: Arc<SqlFlightClient>,
+    sql: &str,
+    params: Option<RecordBatch>,
+) -> RetryableQueryStream {
+    RetryableQueryStream::new(client, sql, params)
+}
+
+/// Represents the current state of the `RetryableQueryStream` state machine.
+///
+/// The stream transitions through these states during query execution:
+/// `Ready` → `Executing` → `Streaming` → `Error` (`is_retryable`: True) → `Ready`
+/// or
+/// `Ready` → `Executing` → `Error` (`is_retryable`: True) → `Ready`
+/// or
+/// `Ready` → `Executing` → `Error` (`is_retryable`: False) -> `Terminated`.
+enum StreamState {
+    /// Initial state, ready to start or retry a query
+    Ready,
+    /// Query is being executed, waiting for the server to return a stream
+    Executing(Pin<Box<dyn Future<Output = Result<FlightRecordBatchStream, GenericError>> + Send>>),
+    /// Actively streaming record batches from the server
+    Streaming(Pin<Box<FlightRecordBatchStream>>),
+    /// A retryable error occurred and has been yielded to the consumer
+    Error {
+        error: GenericError,
+        is_retryable: bool,
+    },
```

Comment:
Why not name this `RetryableError` and remove the boolean. non-retryable errors should just go to the `Terminated` state directly